src/core/utils/HashUtils.ts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2021 SYMBOL
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Hash, keccak256, keccak512, sha3_256 } from 'js-sha3';
+import * as ripemd160 from 'ripemd160';
+
+/**
+ * The type of data HashUtils can hash.
+ */
+type Message = Uint8Array | string;
+
+/**
+ * Interface of a hasher that keeps the state of the hashed data.
+ */
+export interface Hasher {
+    /**
+     * @param data - the data to hash
+     * @returns this hasher
+     */
+    update(data: Message): Hasher;
+
+    /**
+     * returns the hash of the provided data.
+     * @returns the hash of the data
+     */
+    digest(): Uint8Array;
+
+    /**
+     * It resets the hashed data to empty
+     * @returns this hasher
+     */
+    reset(): Hasher;
+}
+
+/**
+ * A utility class to hash data easily. It handles keccak256, keccak512, sha3_256 and ripemd160 implementations.
+ *
+ * For each implementation, this helper provides a Hasher factory and a one-line hash function.
+ *
+ * This class abstracts out the the implementation, if for some reason the hashes need to be changed, it's centralized in one place.
+ *
+ * Clients wouldn't need to import any hash implementation like 'ripemd160' or 'js-sha3', just use this helper utils.
+ */
+export class HashUtils {
+    /**
+     * It keccak512 hashes the list of Uint8Array into a Uint8Array.
+     *
+     * @param data - the list of Uint8Array to hash.
+     * @returns the keccak512 hash of the data.
+     */
+    public static keccak512Hash = (...data: Message[]): Uint8Array => {
```

Comment:
why it need `...data` rather than just `data`?

Reply:
code client can use:
- `keccak512Hash(a)` instead of `keccak512Hash([a])`
- `keccak512Hash(a,b,c,d)` instead of `keccak512Hash([a,b,c,d])`